INTERNALS.md

```diff
@@ -0,0 +1,205 @@
+# `jemalloc` internals
+
+## High-level overview
+
+Any given allocation size is internally increased to round it up to the nearest
+"size class".  These size classes are chosen to balance chances for reuse,
+avoiding overconsumption, and making internal calculations easier.  We divide
+the size classes into two types: small (those less than a few pages) and large.
+The way in which we allocate depends on whether or not the allocation size is
+small or large.  Large allocations are satisfied from a more central page-level
+allocator.  For small allocations, we allocate a "slab" from that page-level
+allocator, of some small number of pages (1-7, given our current size class
+selections and common page sizes).  It is dedicated only to allocations of a
+single size class.  (Every slab is a slab of *some* size class; the
+distinguishing feature of a slab, as opposed to some other reservation of pages,
+is that it is reserved for allocations of a small size class).  The Centralized
+slab and page allocators are sharded into "arenas".
+
+Sitting in front of the arena-level allocators is a thread-local cache
+("tcache").  Allocations up to some particular size are filled from the tcache
+when it is nonempty.  When it is empty, the tcache attempts to fill itself using
+the more centralized mechanism (using items from a slab if the relevant size is
+small, and a page allocation if the size is large).  Deallocated memory is put
+into the tcache if possible (and is flushed to the more central allocator if
+not).
+
+## Some key concepts
+
+### Size classes
+
+There are two relevant modules here:
+- `sc`: Actually sets the size classes (according to static and runtime
+    configuration options).
+- `sz`: Converts in between various notions of sizes (size in bytes, the size
+    class index, or a "page size" index, which is used in making requests from
+    the page-level allocator).
+
+### Extents
+
+An extent is a page-aligned, multiple-of-page-sized region of the virtual
+address space owned by jemalloc.
+
+### Extent states and purging
+
+Modern architectures perform virtual -> physical address translation, and modern
+OSs in turn allow features like overcommit; reserving virtual address space
+without backing it with physical pages.
+
+We say that a page of virtual memory is in one of four states:
+- Active: In use by the application
+- Dirty: No longer in use, but we haven't told the OS yet.
+- Muzzy: We've told the OS to reclaim the page, but asynchronously.  It may
+  never get reclaimed, and definitely won't if we touch it again.
+- Retained: We've told the OS to reclaim the page, and waited synchronously for
+  it to do so.
```

Comment:
Maybe also mention that, if unmapped, we don't keep tracking it anymore? Feel free to ignore.